pkgs/development/python-modules/jaxlib/default.nix

```diff
@@ -1,90 +1,280 @@
+      # directory; not sure why but this works around it
+      "${cudatoolkit}/targets/${stdenv.system}"
+    ];
+  };
+
+  cudatoolkit_cc_joined = symlinkJoin {
+    name = "${cudatoolkit.cc.name}-merged";
+    paths = [
+      cudatoolkit.cc
+      binutils.bintools # for ar, dwp, nm, objcopy, objdump, strip
+    ];
+  };
+
+  bazel-build = buildBazelPackage {
+    name = "bazel-build-${pname}-${version}";
+
+    bazel = bazel_3;
+
+    src = fetchFromGitHub {
+      owner = "google";
+      repo = "jax";
+      rev = "${pname}-v${version}";
+      sha256 = "01ks4djbpjsxjy2zwdwv3h00sgwi4ps3jz75swddrw2f56zjdmw4";
     };
-    gpu = fetchurl {
-      url = "https://storage.googleapis.com/jax-releases/cuda111/jaxlib-${version}+cuda111-cp39-none-manylinux2010_x86_64.whl";
-      sha256 = "sha256:065kyzjsk9m84d138p99iymdiiicm1qz8a3iwxz8rspl43rwrw89";
+
+    nativeBuildInputs = [
+      cython
+      pkgs.flatbuffers
+      git
+      setuptools
+      wheel
+    ];
+
+    buildInputs = [
+      double-conversion
+      giflib
+      grpc
+      jsoncpp
+      libjpeg_turbo
+      pkgs.flatbuffers
+      pkgs.protobuf
+      pybind11
+      snappy
+      zlib
+    ] ++ lib.optionals cudaSupport [
+      cudatoolkit
+      cudnn
+    ];
+
+    propagatedBuildInputs = [
+      numpy
+      scipy
+      six
+    ];
+
+    postPatch = ''
+      rm -f .bazelversion
+    '';
+
+    bazelTarget = "//build:build_wheel";
+
+    removeRulesCC = false;
+
+    GCC_HOST_COMPILER_PREFIX = lib.optionalString cudaSupport "${cudatoolkit_cc_joined}/bin";
+    GCC_HOST_COMPILER_PATH = lib.optionalString cudaSupport "${cudatoolkit_cc_joined}/bin/gcc";
+
+    preConfigure = ''
+      # dummy ldconfig
+      mkdir dummy-ldconfig
+      echo "#!${stdenv.shell}" > dummy-ldconfig/ldconfig
+      chmod +x dummy-ldconfig/ldconfig
+      export PATH="$PWD/dummy-ldconfig:$PATH"
+      cat <<CFG > ./.jax_configure.bazelrc
+      build --strategy=Genrule=standalone
+      build --repo_env PYTHON_BIN_PATH="${python}/bin/python"
+      build --action_env=PYENV_ROOT
+      build --python_path="${python}/bin/python"
+      build --distinct_host_configuration=false
+    '' + lib.optionalString cudaSupport ''
+      build --action_env CUDA_TOOLKIT_PATH="${cudatoolkit_joined}"
+      build --action_env CUDNN_INSTALL_PATH="${cudnn}"
+      build --action_env TF_CUDA_PATHS="${cudatoolkit_joined},${cudnn},${nccl}"
+      build --action_env TF_CUDA_VERSION="${lib.versions.majorMinor cudatoolkit.version}"
+      build --action_env TF_CUDNN_VERSION="${lib.versions.major cudnn.version}"
+      build:cuda --action_env TF_CUDA_COMPUTE_CAPABILITIES="${lib.concatStringsSep "," cudaCapabilities}"
+    '' + ''
+      CFG
+    '';
+
+    # Copy-paste from TF derivation.
+    # Most of these are not really used in jaxlib compilation but it's simpler to keep it
+    # 'as is' so that it's more compatible with TF derivation.
```

Comment:
I'd really prefer to trim this down to only the essentials. But I'm ok with merging it for now and then refactoring later.

Reply:
There are several issues with modifying this list:
* I'm not aware of any easy way of determining which packages exactly are needed for `jaxlib` ouf of this list. One could try to either manually analyse all transitive dependencies from `build/BUILD.bazel` and match them to the `TF_SYSTEM_LIBS` list or write and maintain a (non-trivial) automated script to do so. Just "commenting out one by one and try" approach won't work as the dependencies will be just replaced by Bazel-provided ones.
* Even if the above step is done - this list changes regularly and, therefore, needs to be maintained. Copy-pasting it from `tensorflow` build is much easier than re-doing this analysis on each change.
* The case of missing dependencies on Nix side (= if this list claims some deps are provided by system but are actually missing in Nix build environment) is easy to spot: the build will fail (and typically quite early). The opposite case (if the list is too short / doesn't include some of the possible dependencies) is a 'silent failure': the build will succeed, it's just that Bazel will inject its own provided dependencies. Hence, I'd prefer to keep this list as comprehensive as possible to bump into the build failures instead of silent pulls of deps by Bazel.

Reply:
ack ok I see... this is messier than I realized. i don't understand all the technical details exactly but I see where you're coming from. ideally we could come up with a cleaner way of handling `TF_SYSTEM_LIBS`, but that's def not in scope for this PR